### PR TITLE
PADV-414: LTI launch course permission

### PR DESCRIPTION
**Pull request by kuipumu**
## Ticket

https://agile-jira.pearson.com/browse/PADV-414

## Description

This PR adds a course permission to the LTI launch view. We included this feature to restrict access to courses per LTI tool, this feature can be enabled with the 'openedx_lti_tool_plugin.course_access_configuration' waffle switch. In order to make the course permission to work this includes a CourseAccessConfiguration model, this model stores a list of courses ID allowed for a LTI tool, on the LTI launch view, if the feature is enabled, it will verify the requested course ID exists for the tool configuration, if not, it will return a HTTP 400 error.

## Type of Change

- [x] Add CourseAccessConfiguration middleware class.
- [x] Add COURSE_ACCESS_CONFIGURATION WaffleSwitch.
- [x] Modify LtiToolLaunchView post method.
- [x] Add edx-toggles base.in requirement.
- [x] Upgrade requirements.
- [x] Include tests for all added  and modified code.

## Testing:

- Run the LMS: `make dev.up.lms`.
- Install `openedx_lti_tool_plugin` on the LMS.
- Add required settings to LMS.
	
```
# Enable LTI Tool Provider Plugin.
OLTITP_ENABLE_LTI_TOOL = True
```

- Run ngrok or any other similar service: `ngrok http 18000`
- Create a new LTI 1.3 tool key config: http://localhost:18000/admin/lti1p3_tool_config/ltitoolkey/add/ (You can use IMS RI to create a key pair: https://lti-ri.imsglobal.org/keygen/index)
- Create a new LTI 1.3 tool config: http://localhost:18000/admin/lti1p3_tool_config/ltitool/add/

```
Issuer: https://saltire.lti.app/platform
Client id: saltire.lti.app
Auth login: url: https://saltire.lti.app/platform/auth
Auth token url: https://saltire.lti.app/platform/token/sda42bb0cf2352259a367a404d48d54e8
Key set url: https://saltire.lti.app/platform/jwks/sda42bb0cf2352259a367a404d48d54e8
Deployment ids: ["cLWwj9cbmkSrCNsckEFBmA"]
```

- Create a new CourseAccessConfiguration instance for the previously created LTI tool and include the course_id in the "Allowed Course IDs" list.
- Go to the saLTIre platform https://saltire.lti.app/platform
- Go to "Security Model" on the left sidebar and set these settings:

```
Message URL: https://{ngrok_url}/openedx_lti_tool_plugin/1.3/launch/{course_id}
Initiate login URL: https://{ngrok_url}/openedx_lti_tool_plugin/1.3/login
Redirection URI(s): https://{ngrok_url}/openedx_lti_tool_plugin/1.3/launch/{course_id}
Public keyset URL: https://{ngrok_url}/openedx_lti_tool_plugin/1.3/pub/jwks
```

- Click on the "Save" button on the top navbar.
- Click on the dropdown next to the "Connect" button on the top navbar.
- Click on "Open in iframe".
- The launch should work.
- Repeat the launch with a different Message URL (Also update Redirection URI(s) field to include the new message URL).
- The launch should fail with a HTTP 400 response.

## Reviewers

- [ ] @Squirrel18 
- [ ] @alexjmpb

### CODE REVIEW

File: openedx_lti_tool_plugin/tests/test_models.py [hunk 1]
```diff
@@ -137,3 +139,91 @@ def test_save_method_with_user(
...
+        """Test fixtures setup."""
+        super().setUp()
+        self.lti_tool = LtiTool.objects.create(
+            client_id='random-client-id',
+            auth_login_url='random-login-url',
+            auth_token_url='random-token-url',
+            deployment_ids='["random-deployment-id"]',
+            tool_key=LtiToolKey.objects.create(),
+        )
+        self.allowed_course_ids = ['id-1', 'id-2']
```
Comment (alexjmpb) [inline openedx_lti_tool_plugin/tests/test_models.py:157]: Shouldn't allowed_course_ids be a JSON string?

File: openedx_lti_tool_plugin/tests/test_models.py [hunk 2]
```diff
@@ -137,3 +139,91 @@ def test_save_method_with_user(
...
+    @patch('openedx_lti_tool_plugin.models.json.loads')
+    def test_is_course_id_allowed_with_allowed_course_id(self, json_loads_mock: MagicMock):
+        """Test is_course_id method with allowed course ID.
+
+        Args:
+            json_loads_mock: Mocked json.loads function.
+        """
+        json_loads_mock.return_value = self.allowed_course_ids
+
+        self.assertEqual(self.profile.is_course_id_allowed('id-1'), True)
```
Comment (alexjmpb) [inline openedx_lti_tool_plugin/tests/test_models.py:212]:
```suggestion
        self.assertTrue(self.profile.is_course_id_allowed('id-1'))
```

File: openedx_lti_tool_plugin/tests/test_models.py [hunk 3]
```diff
@@ -137,3 +139,91 @@ def test_save_method_with_user(
...
+    @patch('openedx_lti_tool_plugin.models.json.loads')
+    def test_is_course_id_allowed_with_unknown_course_id(self, json_loads_mock: MagicMock):
+        """Test is_course_id method with unknown course ID.
+
+        Args:
+            json_loads_mock: Mocked json.loads function.
+        """
+        json_loads_mock.return_value = self.allowed_course_ids
+
+        self.assertEqual(self.profile.is_course_id_allowed('id-3'), False)
```
Comment (alexjmpb) [inline openedx_lti_tool_plugin/tests/test_models.py:224]:
```suggestion
        self.assertFalse(self.profile.is_course_id_allowed('id-3'))
```

File: openedx_lti_tool_plugin/waffle.py [hunk 4]
```diff
@@ -0,0 +1,9 @@
+
+"""Waffle flags for openedx_lti_tool_plugin."""
+from edx_toggles.toggles import WaffleSwitch
+
+from openedx_lti_tool_plugin.apps import OpenEdxLtiToolPluginConfig as AppConfig
+
+COURSE_ACCESS_CONFIGURATION = WaffleSwitch(
+    f'{AppConfig.name}.course_access_configuration', __name__
+)
```
Comment (alexjmpb) [inline openedx_lti_tool_plugin/waffle.py:9]:
```suggestion

WAFFLE_NAMESPACE = AppConfig.name

COURSE_ACCESS_CONFIGURATION = WaffleSwitch(
    f'{WAFFLE_NAMESPACE}.course_access_configuration', __name__
)
```

What do you think of adding this constant to be more explicit?

File: openedx_lti_tool_plugin/tests/test_views.py [hunk 5]
```diff
@@ -312,48 +312,235 @@ def test_enroll_without_enrollment(self, course_enrollment_mock: MagicMock):
...
+            tool_storage_mock: Mocked DjangoCacheDataStorage class.
+            get_launch_data_mock: Mocked DjangoMessageLaunch get_launch_data method.
+            course_access_configuration_mock: Mocked COURSE_ACCESS_CONFIGURATION object.
+            course_access_configuration_filter_mock: Mocked CourseAccessConfiguration.ojects filter method.
+            authenticate_and_login_mock: Mocked LtiToolLaunchView authenticate_and_login method.
+            course_key_mock: Mocked CourseKey from_string method.
+            enroll_mock: Mocked enroll LtiToolLaunchView method.
+            redirect_mock: Mocked redirect function.
+        """
+        course_access_configuration_mock.is_enabled.return_value = True
```
Comment (alexjmpb) [inline openedx_lti_tool_plugin/tests/test_views.py:350]: What do you think of using this decorator https://github.com/openedx/edx-toggles/blob/master/edx_toggles/toggles/testutils.py#L7 to mock the value of the waffle flag?

File: openedx_lti_tool_plugin/tests/test_views.py [hunk 6]
```diff
@@ -312,48 +312,235 @@ def test_enroll_without_enrollment(self, course_enrollment_mock: MagicMock):
...
+        response = self.view_class.as_view()(request, COURSE_ID)
+
+        message_launch_mock.assert_called_once_with(request, tool_conf_mock(), launch_data_storage=tool_storage_mock())
+        get_launch_data_mock.assert_called_once_with()
+        course_access_configuration_mock.is_enabled.assert_called_once_with()
+        course_access_configuration_filter_mock.assert_called_once_with(
+            lti_tool=tool_conf_mock().get_lti_tool(ISS, AUD),
+        )
+        course_access_configuration_filter_mock().first.assert_called_once_with()
+        course_access_configuration_filter_mock().first().is_course_id_allowed.assert_called_once_with(COURSE_ID)
```
Comment (alexjmpb) [inline openedx_lti_tool_plugin/tests/test_views.py:363]:
```suggestion
        course_access_configuration_filter_mock().first.assert_called_once_with()
course_access_configuration_filter_mock().first().is_course_id_allowed.assert_called_once_with(COURSE_ID)
```

File: openedx_lti_tool_plugin/tests/test_views.py [hunk 7]
```diff
@@ -312,48 +312,235 @@ def test_enroll_without_enrollment(self, course_enrollment_mock: MagicMock):
...
             get_launch_data_mock: Mocked DjangoMessageLaunch get_launch_data method.
-            course_key_mock: Mocked CourseKey from_string method.
+            course_access_configuration_mock: Mocked COURSE_ACCESS_CONFIGURATION object.
+            course_access_configuration_filter_mock: Mocked CourseAccessConfiguration.ojects filter method.
             authenticate_and_login_mock: Mocked LtiToolLaunchView authenticate_and_login method.
+            course_key_mock: Mocked CourseKey from_string method.
             enroll_mock: Mocked enroll LtiToolLaunchView method.
             redirect_mock: Mocked redirect function.
         """
+        course_access_configuration_mock.is_enabled.return_value = False
```
Comment (alexjmpb) [inline openedx_lti_tool_plugin/tests/test_views.py:411]: Same comment about using https://github.com/openedx/edx-toggles/blob/master/edx_toggles/toggles/testutils.py#L7

File: openedx_lti_tool_plugin/tests/test_views.py [hunk 8]
```diff
@@ -312,48 +312,235 @@ def test_enroll_without_enrollment(self, course_enrollment_mock: MagicMock):
...
+        Args:
+            message_launch_mock: Mocked DjangoMessageLaunch class.
+            tool_conf_mock: Mocked DjangoDbToolConf class.
+            tool_storage_mock: Mocked DjangoCacheDataStorage class.
+            get_launch_data_mock: Mocked DjangoMessageLaunch get_launch_data method.
+            course_access_configuration_mock: Mocked COURSE_ACCESS_CONFIGURATION object.
+            course_access_configuration_filter_mock: Mocked CourseAccessConfiguration.ojects filter method.
+            log: LogCapture fixture.
+        """
+        course_access_configuration_mock.is_enabled.return_value = True
```
Comment (alexjmpb) [inline openedx_lti_tool_plugin/tests/test_views.py:458]: Same comment about using https://github.com/openedx/edx-toggles/blob/master/edx_toggles/toggles/testutils.py#L7

File: openedx_lti_tool_plugin/tests/test_views.py [hunk 9]
```diff
@@ -312,48 +312,235 @@ def test_enroll_without_enrollment(self, course_enrollment_mock: MagicMock):
...
```
Comment (alexjmpb) [inline openedx_lti_tool_plugin/tests/test_views.py:514]: Same comment about using https://github.com/openedx/edx-toggles/blob/master/edx_toggles/toggles/testutils.py#L7

File: openedx_lti_tool_plugin/models.py [hunk 10]
```diff
@@ -124,3 +129,74 @@ def save(self, *args: tuple, **kwargs: dict):
...
+        LtiTool,
+        on_delete=models.CASCADE,
+        related_name='openedx_lti_tool_plugin_course_permission',
+        verbose_name=_('LTI Tool'),
+    )
+    allowed_course_ids = models.TextField(
+        verbose_name=_('Allowed Course IDs'),
+        help_text=_(f'List of allowed courses IDs. {EXAMPLE_ID_LIST}'),
+        default=[],
+    )
```
Comment (Jacatove) [inline openedx_lti_tool_plugin/models.py:152]: Is this field going to be used only trough the admin panel?. If yes, what do you think about including a course id validator?.
Didn't you consider making a many to many relationship to courses? I suppose this is not going to be friendly in admin panel.

@kuipumu

File: openedx_lti_tool_plugin/signals.py [hunk 11]
```diff
@@ -0,0 +1,26 @@
+"""Signals for openedx_lti_tool_plugin."""
+from django.db.models.signals import post_save
+from django.dispatch import receiver
+from pylti1p3.contrib.django.lti1p3_tool_config.models import LtiTool
+
+from openedx_lti_tool_plugin.models import CourseAccessConfiguration
+
+
+@receiver(post_save, sender=LtiTool, dispatch_uid='create_access_configuration_on_lti_tool_create')
```
Comment (Squirrel18) [inline openedx_lti_tool_plugin/signals.py:9]:
```suggestion
@receiver(post_save, sender=LtiTool, dispatch_uid='create_access_configuration_on_lti_tool_creation')
```

File: openedx_lti_tool_plugin/tests/test_views.py [hunk 12]
```diff
@@ -315,32 +316,35 @@ def test_enroll_without_enrollment(self, course_enrollment_mock: MagicMock):
...
     @patch.object(CourseKey, 'from_string', return_value='random-course-key')
-    @patch.object(DjangoMessageLaunch, 'get_launch_data', return_value=BASE_LAUNCH_DATA)
+    @patch.object(LtiToolLaunchView, 'authenticate_and_login', return_value='random-user')
+    @patch.object(CourseAccessConfiguration.objects, 'filter')
+    @patch.object(DjangoMessageLaunch, 'get_launch_data', return_value={**BASE_LAUNCH_DATA, 'azp': AUD})
     @patch('openedx_lti_tool_plugin.views.DjangoCacheDataStorage')
     @patch('openedx_lti_tool_plugin.views.DjangoDbToolConf')
     @patch.object(DjangoMessageLaunch, '__init__', return_value=None)
-    def test_post_with_identity_claims(
+    def test_post_with_course_launch(
```
Comment (Squirrel18) [inline openedx_lti_tool_plugin/tests/test_views.py:326]: Please, check that CourseAccessConfiguration.objects.filter and self.tool_config.get_lti_tool have been called with the correct parameters. @kuipumu

**Review by Squirrel18 — CHANGES_REQUESTED** [review]

**Review by Squirrel18 — APPROVED** [review]

**Pull request merged**